File: mobile/src/types/assistant.ts
/**
 * Assistant 领域类型 — Stage 4 共享契约
 * 与后端 models/assistant.py 一一对应（camelCase，populate_by_name=True）
 */

// ── 会话消息 ──────────────────────────────────────

export type AssistantRole = 'user' | 'assistant';

export interface AssistantHistoryMessage {
  role: AssistantRole;
  text: string;
}

// ── 产品建议 ──────────────────────────────────────

export type AssistantRecommendation =
  | 'recommended'
  | 'not_recommended'
  | 'needs_information';

export interface AssistantProductAdvice {
  productId: string;
  productName: string;
  recommendation: AssistantRecommendation;
  reason: string;
  steps: string[];
  cautions: string[];
}

// ── 安全警告 ──────────────────────────────────────

export type AssistantWarningSeverity = 'critical' | 'attention' | 'unknown';

export interface AssistantSafetyWarning {
  severity: AssistantWarningSeverity;
  title: string;
  description: string;
  relationId: string | null;
  recommendedAction: string;
}

// ── 来源引用（与后端 SourceRef 对齐，camelCase）──

export type SourceRefType = 'local_kb' | 'warehouse' | 'rule' | 'external';

export interface SourceRef {
  type: SourceRefType;
  title: string;
  domain: string;
  url: string;
  retrievedAt: string | null;
  ref: string;
  version: string;
}

// ── 知识证据（仅 reviewed 条目，与后端 KnowledgeEvidence 对齐）──

export interface KnowledgeEvidence {
  entryId: string;
  topic: string;
  surfaces: string[];
  excludedSurfaces: string[];
  steps: string[];
  warnings: string[];
  prohibitedActions: string[];
  stopConditions: string[];
  tagCondition: string;
  sources: SourceRef[];
  confidence: 'reviewed';
}

// ── 知识状态 ──────────────────────────────────────

export type KnowledgeStatus =
  | 'local_hit'
  | 'local_hit_no_inventory'
  | 'external_hit'
  | 'insufficient'
  | 'no_match'
  | 'external_fail';

// ── 最终响应 ──────────────────────────────────────

/**
 * 助手响应。Stage 3 新增字段在旧后端缺失时可选，运行时由
 * normalizeAssistantResponse 补齐安全默认值，避免崩溃。
 */
export interface AssistantResponse {
  answer: string;
  needsClarification: boolean;
  clarificationQuestions: string[];
  inventoryAdvice: AssistantProductAdvice[];
  generalAdvice: string[];
  safetyWarnings: AssistantSafetyWarning[];
  outOfScope: boolean;
  evidence: string[];
  // ── Stage 3 知识库优先（旧后端缺失时安全降级）──
  knowledge?: KnowledgeEvidence[];
  knowledgeStatus?: KnowledgeStatus;
  externalSources?: SourceRef[];
  sources?: SourceRef[];
  pendingKnowledgeNotice?: string;
}

// ── 本地会话消息（含 UI 状态）────────────────────

export interface AssistantMessage {
  id: string;
  role: AssistantRole;
  text: string;
  imageUri?: string;
  response?: AssistantResponse;
}

File: backend/app/models/assistant.py
"""助手问答领域模型 — Stage 4 共享契约

productId 必须来自数据库，不允许由 LLM 只返回产品名称。
所有模型开启 populate_by_name=True，保持移动端 camelCase 契约。
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.models.knowledge import (
    KnowledgeEvidence,
    KnowledgeStatus,
    SourceRef,
)

# ── 会话消息 ──────────────────────────────────────

class AssistantHistoryMessage(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    role: Literal["user", "assistant"]
    text: str = Field(max_length=4000)

# ── 产品建议 ──────────────────────────────────────

class AssistantProductAdvice(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    product_id: str = Field(..., alias="productId")
    # 产品名必须来自数据库，不允许由 LLM 返回；后端 _assemble 阶段由库存回填
    product_name: str = Field(default="", alias="productName")
    recommendation: Literal["recommended", "not_recommended", "needs_information"]
    reason: str
    steps: list[str] = Field(default_factory=list)
    cautions: list[str] = Field(default_factory=list)

# ── 安全警告 ──────────────────────────────────────

class AssistantSafetyWarning(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    severity: Literal["critical", "attention", "unknown"]
    title: str
    description: str
    relation_id: str | None = Field(None, alias="relationId")
    rule_id: str | None = Field(None, alias="ruleId")
    recommended_action: str

# ── 最终响应 ──────────────────────────────────────

class AssistantResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    answer: str
    needs_clarification: bool = Field(alias="needsClarification")
    clarification_questions: list[str] = Field(
        default_factory=list, alias="clarificationQuestions"
    )
    inventory_advice: list[AssistantProductAdvice] = Field(
        default_factory=list, alias="inventoryAdvice"
    )
    general_advice: list[str] = Field(default_factory=list, alias="generalAdvice")
    safety_warnings: list[AssistantSafetyWarning] = Field(
        default_factory=list, alias="safetyWarnings"
    )
    out_of_scope: bool = Field(False, alias="outOfScope")
    evidence: list[str] = Field(default_factory=list)
    # ── 知识库优先（Stage 3）新增 ──
    knowledge: list[KnowledgeEvidence] = Field(default_factory=list, alias="knowledge")
    knowledge_status: KnowledgeStatus = Field("no_match", alias="knowledgeStatus")
    external_sources: list[SourceRef] = Field(default_factory=list, alias="externalSources")
    sources: list[SourceRef] = Field(default_factory=list)
    pending_knowledge_notice: str = Field(default="", alias="pendingKnowledgeNotice")

# ── Legacy AI 候选草稿（已废弃，新知识库优先流程不调用）──

class LegacyAssistantDraft(BaseModel):
    """[legacy] 旧版单次 LLM 输出草稿。

    仅被旧接口 answer_household_question 使用；新问答流程不使用本模型，
    也不会用其 product_advice/safety_warnings/out_of_scope 覆盖 AssistantResponse。
    新流程只使用 KnowledgeIntentDraft 与 AssistantNarrativeDraft。
    """

    model_config = ConfigDict(populate_by_name=True)

    answer: str = ""
    needs_clarification: bool = False
    clarification_questions: list[str] = Field(default_factory=list)
    product_advice: list[AssistantProductAdvice] = Field(default_factory=list)
    general_advice: list[str] = Field(default_factory=list)
    safety_warnings: list[AssistantSafetyWarning] = Field(default_factory=list)
    out_of_scope: bool = False
